overlord/fdestate/dbx.go

```diff
@@ -0,0 +1,568 @@
+	if err != nil {
+		return err
+	}
+
+	if op == nil {
+		logger.Debugf("no pending DBX update request")
+		return nil
+	}
+
+	// we have a pending request, which means we must mark it as failed and
+	// reseal with the current content of EFI DBX
+
+	if op.Status != DoingStatus {
+		return &snapstate.ChangeConflictError{
+			ChangeKind: "fde-efi-secureboot-db-update",
+			Message:    "cannot perform DBX update 'startup' action when conflicting actions are in progress",
+		}
+	}
+
+	op.SetStatus(AbortingStatus)
+	op.Err = "'startup' action invoked while an operation is in progress"
+	if err := updateExternalOperation(st, op); err != nil {
+		return nil
+	}
+
+	chg := st.Change(op.ChangeID)
+	return completeEFISecurebootDBUpdateChange(chg)
+}
+
+type dbxUpdateContext struct {
+	Payload []byte               `json:"payload"`
+	Method  device.SealingMethod `json:"sealing-method"`
+}
+
+// addEFISecurebootDBUpdateChange adds a state change related to the DBX
+// update. The state must be locked by the caller
+func addEFISecurebootDBUpdateChange(st *state.State, method device.SealingMethod, payload []byte) (*externalOperation, error) {
+	// add a change carrying 2 tasks:
+	// - efi-secureboot-db-update-prepare: with a noop do, but the undo handler
+	// preforms necessary cleanup
+	// - efi-secureboot-db-update: waits for the external caller to indicate
+	// that the action is complete or failed
+	//
+	// if the original requester never calls cleanup/startup, the change
+	// will be pruned automatically and the undo will perform a reseal
+
+	tPrep := st.NewTask("efi-secureboot-db-update-prepare", "Prepare for external EFI DBX update")
+	tUpdateWait := st.NewTask("efi-secureboot-db-update", "External EFI DBX update")
+	tUpdateWait.WaitFor(tPrep)
+	ts := state.NewTaskSet(tPrep, tUpdateWait)
+
+	chg := st.NewChange("fde-efi-secureboot-db-update", "EFI DBX update")
+	chg.AddAll(ts)
+
+	data, err := json.Marshal(dbxUpdateContext{
+		Payload: payload,
+		Method:  method,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	op := &externalOperation{
+		// match the change kind
+		Kind:     "fde-efi-secureboot-db-update",
+		ChangeID: chg.ID(),
+		Context:  json.RawMessage(data),
+		Status:   PreparingStatus,
+	}
+
+	err = addExternalOperation(st, op)
+	if err != nil {
+		return nil, err
+	}
+
+	var syncChs map[string]chan struct{}
+	val := st.Cached(dbxUpdatePrepareSyncKey{})
+	if val == nil {
+		syncChs = make(map[string]chan struct{})
+	} else {
+		syncChs = val.(map[string]chan struct{})
+	}
+	syncChs[chg.ID()] = make(chan struct{})
+	st.Cache(dbxUpdatePrepareSyncKey{}, syncChs)
+
+	return op, nil
+}
+
+func completeEFISecurebootDBUpdateChange(chg *state.Change) error {
+	st := chg.State()
+
+	// trigger ensure so that the task runner attempts to run our tasks
+	st.EnsureBefore(0)
+
+	st.Unlock()
+	logger.Debugf("waiting for FDE DBX change %v to become ready", chg.ID())
+	<-chg.Ready()
```

Comment:
do we need a timeout here?

Reply:
The change is supposed to complete at this point, we're not holding the state lock and this is running through daemon API so hitting a timeout would result in uncertainty about the overall status of the change.

The caller also does not know the change ID, so they would be unable to retry the request if the status of the external operation has been change to something else than Doing.